### ISSUE TITLE
Support for MPI?

### ISSUE THREAD

**Issue by darafferty** [DataBiosphere/toil#3517]
Hello,

We are currently using Toil to run CWL workflows on a Slurm cluster. This is working well in general, but we have some jobs that can only use MPI to distribute the work over multiple nodes. I believe that MPI support was added recently to CWL (see https://github.com/common-workflow-language/cwltool/pull/1276), but I couldn't find any support for it in Toil currently -- is that correct? If so, are there any plans to add support, specifically for use with CWL+Slurm?

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-838)
┆Issue Number: TOIL-838

**Comment by DailyDreaming** [comment]
@darafferty Looking at that PR, it looks like Toil might work with MPI if we just add the argument, since since we inherit cwltool's run methods.

I've made a PR, would you mind trying it out?  :)

https://github.com/DataBiosphere/toil/pull/3525

**Comment by DailyDreaming** [comment]
@darafferty This should be in master now.  Please try it out and let us know how it goes!

**Comment by adamnovak** [comment]
We think this does in fact work.

**Comment by darafferty** [comment]
Thanks again for implementing this, and sorry for my long delay in testing it. I have now tried it, and I'm finding that the values in the MPI configuration file (passed with the `--mpi-config-file` arg) seem to be ignored and the defaults used instead. However, it works if I add the following lines to `cwltoil.py` (around line 2195):

```
from cwltool.mpi import MpiConfig
runtime_context.mpi_config = MpiConfig.load(options.mpi_config_file)

```
This probably isn't the right/best way to do it, but at any rate I can confirm that with it I'm able to run MPI jobs using SLURM!

**Comment by mr-c** [comment]
Thank you for the report @darafferty !

Here is the equivalent code from `cwltool`

https://github.com/common-workflow-language/cwltool/blob/e37134a90ac7a7c18254e30cff16da590b45c6d7/cwltool/main.py#L950

``` python
from .mpi import MpiConfig

# […]

        if args.mpi_config_file is not None:
            runtimeContext.mpi_config = MpiConfig.load(args.mpi_config_file)
```

A PR to add this to cwltoil.py would be appreciated!

**Comment by darafferty** [comment]
I created a PR for the fix (see https://github.com/DataBiosphere/toil/pull/3574). One issue that relates specifically to using MPI with SLURM, at least for the way I've gotten it to work, is that an extra node is used beyond those used for the MPI job itself. Basically, I use `salloc` as the 'runner' in the MPI config file. This means that Toil calls `salloc` to request a Slurm allocation and run `mpirun`, but the node on which `salloc` is run is not used for the MPI job and is mostly idle as a result. It does work fine, but I wonder if there is a better way to use MPI and Slurm with Toil?

**Comment by mr-c** [comment]
> I created a PR for the fix (see #3574)

Thanks!

> One issue that relates specifically to using MPI with SLURM, at least for the way I've gotten it to work, is that an extra node is used beyond those used for the MPI job itself. Basically, I use `salloc` as the 'runner' in the MPI config file. This means that Toil calls `salloc` to request a Slurm allocation and run `mpirun`, but the node on which `salloc` is run is not used for the MPI job and is mostly idle as a result. It does work fine, but I wonder if there is a better way to use MPI and Slurm with Toil?

Great feedback, thank you. This may show us how to further improve the experimental ` http://commonwl.org/cwltool#MPIRequirement` ; for `toil-cwl-runner` we can tag the MPI scheduling job to run on the leader (instead of itself also being scheduled) by making a new entry in the `CWL_INTERNAL_JOBS` list.

What do you think @rupertnash  ?

**Comment by rupertnash** [comment]
> I created a PR for the fix (see #3574). One issue that relates specifically to using MPI with SLURM, at least for the way I've gotten it to work, is that an extra node is used beyond those used for the MPI job itself. Basically, I use `salloc` as the 'runner' in the MPI config file. This means that Toil calls `salloc` to request a Slurm allocation and run `mpirun`, but the node on which `salloc` is run is not used for the MPI job and is mostly idle as a result. It does work fine, but I wonder if there is a better way to use MPI and Slurm with Toil?

In my use of MPI + cwltool + SLURM I have a standard job script that I submit with `sbatch` and in the script I run `cwltool` directly, hence CWL tool runs on the leader node of my allocation. I've configured cwltool to use `srun` for MPI jobs. My MPI config file is:
```yaml
runner: srun
nproc_flag: -n
env_pass_regex: ["SLURM_.*"]
```

Afraid my SLURM knowledge isn't great...

**Comment by DailyDreaming** [comment]
#3574 is merged.  This should be fixed.
